Code review

phil9987 opened the pull request:
Looking forward to your critical assessment, Kevin :)

lorenz-openai committed:
add conda env creation file

lorenz-openai committed:
update README for env setup

lorenzkuhn committed:
update readme for environment updates

lorenz-openai committed:
add boilerplate flask file upload

lorenz-openai committed:
add boilerplate flask file upload

lorenz-openai committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenz-openai committed:
add secret key to app to enable sessions

lorenz-openai committed:
add local backend mvp

lorenz-openai committed:
add gitignore

lorenz-openai committed:
add uploads folder

phil9987 committed:
created env requirements for linux

lorenz-openai committed:
add Dockerfile template

lorenz-openai committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenz-openai committed:
initial commit of dockerisation

phil9987 committed:
next level UI

phil9987 committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenz-openai committed:
fix Dockerfile configuration

lorenz-openai committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
create uploads and response folders if they don't exist yet

lorenz-openai committed:
actually fix dockerfile

lorenzkuhn committed:
first running version on DO

lorenzkuhn committed:
use port 80 instead of 2375

phil9987 committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenzkuhn committed:
remove conda env calls from Dockerfile

phil9987 committed:
add stylesheet

phil9987 committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

kklein committed:
Implement gan structure and identity training.

kklein committed:
Implement inference based on image file.

phil9987 committed:
improved javascript error handling and brought style of website to next level

phil9987 committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
Add instructions for docker deployment

kklein committed:
Implement bash script to download datasets.

kklein committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

kklein committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
remove old access point

phil9987 committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
handle logging with gunicorn

phil9987 committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
Add docker debug info

phil9987 committed:
increase timeout of gunicorn workers

phil9987 committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
increase workers, increase timeout

phil9987 committed:
enable log to file

phil9987 committed:
make log to file of gunicorn better

kklein committed:
Remove build information from environment.
Allows more systems to recreate package composition, even if not exactly so.

kklein committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
fix visual error in docker debugging section

lorenzkuhn committed:
try to run conda env in boot.sh

kklein committed:
Add image saving of predicted images.

lorenzkuhn committed:
move to use of conda env in docker container

lorenzkuhn committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenzkuhn committed:
clean up docker file

lorenz-openai committed:
create mac version of environment.yml

lorenz-openai committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenzkuhn committed:
created combined requirements file onenv

phil9987 committed:
add combined requirements file to dockerfile

kklein committed:
Make inference work with gpu models.

kklein committed:
Enable gpu usage of gan.

kklein committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenzkuhn committed:
make it work

lorenzkuhn committed:
omit build versions in yml file

lorenzkuhn committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

kklein committed:
Extract modules to separate file.

kklein committed:
mend

phil9987 committed:
added google analytics id

phil9987 committed:
upgrade pytorch to 1.0.1

phil9987 committed:
first version with inference

phil9987 committed:
add support for inference in Dockerfile and sample model

lorenz-openai committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
convert images to rgb before inference

lorenzkuhn committed:
new model

lorenzkuhn committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenz-openai committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenzkuhn committed:
add gpu model and move learning into src

lorenzkuhn committed:
Remove redundant environment files

phil9987 committed:
minor refactoring

lorenzkuhn committed:
Clean up Dockerfile and boot.sh

phil9987 committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
minor refactoring of html file

lorenz-openai committed:
Clean up application_server.py

lorenz-openai committed:
Resolve merge conflicts

lorenz-openai committed:
Fix flask import

lorenz-openai committed:
Apply pylint feedback for application_server.py

lorenzkuhn committed:
Remove redundant learning folder

phil9987 committed:
delete models from git

lorenzkuhn committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

kklein committed:
Remove examplary code.

kklein committed:
Implement cyclegan.

kklein committed:
Merge conflict.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+            # Tensor, work in progress!
+            data = prediction.data.numpy()            
+            new_img = transforms.ToPILImage(mode='RGB')(data)
+            np_image = np.squeeze(prediction.data.numpy(), axis=0)
```
Remove trailing whitespace.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+            app.logger.info('created image {}'.format(img))
+            prediction = model(transform(img).unsqueeze(0))
+            filename_pred = 'prediction_{}.png'.format(filename.split('.')[0])
+            save_image(prediction, 
```
Remove trailing whitespace.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+from flask import Flask, flash, request, redirect, render_template,\
+                  send_from_directory
+from werkzeug.utils import secure_filename
+from gan import Generator
```
Fix import.

lorenzkuhn commented on src/learning/environment.yml:
```diff
@@ -0,0 +1,83 @@
+  - zstd=1.3.7
+  - pip:
+    - torch==0.4.1
+prefix: /home/kevin/.conda/envs/cycleganlearning
```
It seems like the prefix might not be needed when building the environment: https://github.com/conda/conda/issues/4339.
```suggestion

```

kklein replied:
This file should not exist in the first place.
Deleted.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    model.load_state_dict(torch.load(model_path, map_location=device))
+    image_size = (256, 256)
+    transform = transforms.Compose([
```
Reuse method from utils.py.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+transform = None
+
+
+def init_inference(model_path):
```
Could use more descriptive naming.

phil9987 commented on src/learning/infer.py:
```diff
@@ -0,0 +1,31 @@
+
+    for i in range(N_VALIDATION_IMAGES):
+        prediction = model(validation_data[i][0].unsqueeze(0))
+        save_image(prediction, 'prediction%d.png' % i, normalize=True)
```
use string.format() instead of '%' as this is a deprecated string formatting method (https://docs.python.org/release/3.1.5/library/stdtypes.html#old-string-formatting-operations)
suggestion: 'prediction{}.png'.format(i)

kklein replied:
Good point! Since 3.6 f-strings are even more avant-garde and recommended.
Done.

lorenzkuhn commented on src/learning/cyclegan.py:
```diff
@@ -0,0 +1,96 @@
+    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    torch.manual_seed(7)
+    n_residual_blocks = 9
+    use_dropout = False
```
We'll probably want to use ArgumentParser, to pass these parameters as command line arguments. Can be done like this: https://github.com/pytorch/examples/blob/master/snli/util.py

kklein replied:
Most definitely. My thoughts were that we should postpone this to a point in time in which we are at least somewhat confident we can estimate what parameters might be of interest to be adapted.
I figured this point in time was not now. Do you disagree?

phil9987 commented on src/learning/infer.py:
```diff
@@ -0,0 +1,31 @@
+import torch
+import torchvision
+from torchvision.utils import save_image
+import utils
```
only import get_transform, as it is the only thing you are using from utils?

kklein replied:
Done.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+app.config['MAX_CONTENT_LENGTH'] = 50 * 1024 * 1024
+app.config['UPLOAD_FOLDER'] = UPLOAD_FOLDER
+app.config['RESPONSE_FOLDER'] = RESPONSE_FOLDER
+model = Generator(9)
```
Fix instantiation of Generator.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+    ])
+
+
+def allowed_file(filename):
```
Consider using 'is' in naming to express boolean outcome. This function name could also indicate a transformation to an allowed filed type.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+    """
+
+    if request.method == 'POST':
+        # check if the post request has the file part
```
Redundant comment.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+        rcvd_file = request.files['file']
+        # if user does not select file, browser also
+        # submit an empty part without filename
+        if rcvd_file.filename == '':
```
Could either rcvd_file or rcvd_file.filename be 'None'?

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+    :return:
+    """
+
+    if request.method == 'POST':
```
All of the code is in this conditional. Rather not do anything if conditional is unfulfilled.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+                       normalize=True)
+
+            '''
+            # attempt to create image object from output torch.
```
Remove commented code.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+            img = img.convert('RGB')
+            app.logger.info('created image {}'.format(img))
+            prediction = model(transform(img).unsqueeze(0))
+            filename_pred = 'prediction_{}.png'.format(filename.split('.')[0])
```
Could the filename contain several dots and hence a part of the file name be ignored due to this?
If so, this could lead to a problem in scenarios in which we assume conservation.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+
+        # TODO Lorenz: Cover case where file type is not allowed
+        # TODO Lorenz: Enforce max file size limitation
+        if rcvd_file and allowed_file(rcvd_file.filename):
```
Can allowed_file be called on the filename before it being converted to a _secure_ filename?

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+            app.logger.info("received file {}".format(uploaded_file_path))
+            img = Image.open(rcvd_file)
+            img = img.convert('RGB')
+            app.logger.info('created image {}'.format(img))
```
Is this logging the filename of the image? If so, consider making this explicit in code.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+Module defining our flask application server.
+"""
+
+import os
```
Fix import order.

lorenzkuhn commented on src/learning/infer.py:
```diff
@@ -0,0 +1,31 @@
+    model.load_state_dict(torch.load(MODEL_PATH, map_location=device))
+    transform = utils.get_transform(IMAGE_SIZE)
+    validation_data = torchvision.datasets.ImageFolder(
+        root=VALIDATION_DATA_PATH, transform=transform)
```
Maybe I'm missing something, but it's crucial to set the model to evaluation mode. Doing so turns off Dropout and affects the behaviour of BatchNorm, for instance: https://pytorch.org/docs/master/nn.html#torch.nn.BatchNorm1d, https://pytorch.org/docs/stable/nn.html?highlight=eval#torch.nn.Module.eval, https://discuss.pytorch.org/t/what-does-model-eval-do-for-batchnorm-layer/7146
```suggestion
        root=VALIDATION_DATA_PATH, transform=transform)
        model.eval()
```

kklein replied:
Awesome!

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+    gunicorn_logger = logging.getLogger('gunicorn.error')
+    app.logger.handlers = gunicorn_logger.handlers
+    app.logger.setLevel(gunicorn_logger.level)
+    init_inference('gpu_model')
```
Move to global constant.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+ALLOWED_EXTENSIONS = {'bmp', 'png', 'jpg', 'jpeg', 'ppm', 'pgm', 'tif'}
+
+app = Flask(__name__)
+app.secret_key = b'MBWUdbxX;>]vrTL'
```
Does this 'secret' key belong in a public repository?

lorenzkuhn replied:
Definitely not. I'll comment it out for now since we would only need this to have Sessions. If we actually need it, generate a new key and store it in a file which we don't publish.

phil9987 replied:
or let's just delete the line if we don't need it?

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+from gan import Generator
+from torchvision.utils import save_image
+
+UPLOAD_FOLDER = Path.cwd() / 'uploads/'
```
What do you think of instantiating an object containing all of the global state on server start and making said object globally available?
This could then be referenced in upload_file.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+
+app = Flask(__name__)
+app.secret_key = b'MBWUdbxX;>]vrTL'
+app.config['MAX_CONTENT_LENGTH'] = 50 * 1024 * 1024
```
What's the origin of those constants?

lorenzkuhn replied:
This corresponds to a randomly set limit of 50 megabytes.

lorenzkuhn commented on src/learning/modules.py:
```diff
@@ -0,0 +1,113 @@
+import torch.nn as nn
+import utils
+
```
Not very important but maybe we could rename this file to something like nn_modules.py? Might be more explicit since there are also Python modules and the new name would make it clear what the file contains.

kklein replied:
Done.

kklein commented on src/backend/application_server.py:
```diff
@@ -0,0 +1,133 @@
+           filename.rsplit('.', 1)[1].lower() in ALLOWED_EXTENSIONS
+
+
+@app.route('/', methods=['GET', 'POST'])
```
Is the 'GET' behaviour defined implicitly/by the framework or has it been forgotten?
Can't find any code referring to it.

kklein commented on README.md:
````diff
@@ -1 +1,34 @@
+
+In the folder of the flask app:
+```
+$ export FLASK_APP=hello.py
````
Is this up to date?

kklein commented on src/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+# TODO Lorenz: Copy entire backend/static folder instead of individual files
+COPY backend/static/styles/style.css static/styles/style.css
+COPY backend/static/favicon.ico static/favicon.ico
+COPY onenv_requirements.yml environment.yml
```
Why is this renamed?

kklein commented on src/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+COPY backend/static/favicon.ico static/favicon.ico
+COPY onenv_requirements.yml environment.yml
+COPY learning/gpu_model gpu_model
+COPY learning/gan.py gan.py
```
Out of date.

kklein commented on src/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+COPY backend/static/styles/style.css static/styles/style.css
+COPY backend/static/favicon.ico static/favicon.ico
+COPY onenv_requirements.yml environment.yml
+COPY learning/gpu_model gpu_model
```
Can this be done in a more programmatic or dynamic fashion?

kklein commented on src/backend/static/styles/style.css:
```diff
@@ -0,0 +1,74 @@
+}
+
+body {
+  width: 800px;
```
Fix order.

kklein commented on src/backend/static/styles/style.css:
```diff
@@ -0,0 +1,74 @@
+
+h1 {
+  /*font-family: 'Cherry Swash', cursive;*/
+  margin: 0;
```
Fix order.

kklein commented on src/backend/static/styles/style.css:
```diff
@@ -0,0 +1,74 @@
+/* block styling */
+
+header, left, right, footer {
+  padding: 20px;
```
Fix order.

kklein commented on src/backend/static/styles/style.css:
```diff
@@ -0,0 +1,74 @@
+}
+
+left {
+  border-right: 2px solid rgb(100, 98, 98);
```
Fix order.

kklein commented on src/backend/static/styles/style.css:
```diff
@@ -0,0 +1,74 @@
+}
+
+right {
+  display: flex;
```
Fix order.

lorenzkuhn commented on src/learning/modules.py:
```diff
@@ -0,0 +1,113 @@
+    def __init__(self, n_channels, kernel_size, use_dropout):
+        super(ResidualBlock, self).__init__()
+        block_1 = [nn.ReflectionPad2d(1),
+                   nn.Conv2d(n_channels, n_channels, kernel_size, padding=0),
```
When we have a Conv2d or Linear layer followed by a InstanceNorm or BatchNorm layer, we can set the bias of the Conv layer to false, since the bias term is taken care of in the normalisation layer: https://github.com/KaimingHe/deep-residual-networks/issues/10#issuecomment-194037195.
Will save us some parameters.
```suggestion
                   nn.Conv2d(n_channels, n_channels, kernel_size, padding=0, bias=False),
```

kklein replied:
Sweet! :)

kklein commented on src/backend/static/styles/style.css:
```diff
@@ -0,0 +1,74 @@
+}
+
+input {
+  display: block;
```
Fix order.

kklein commented on src/backend/static/styles/style.css:
```diff
@@ -0,0 +1,74 @@
+}
+
+section img {
+  background-color: rgba(126, 126, 126, 0.336);
```
Fix order.

kklein commented on src/backend/static/styles/style.css:
```diff
@@ -0,0 +1,74 @@
+section img {
+  background-color: rgba(126, 126, 126, 0.336);
+  width: 100%;
+  bottom: 0px;
```
Properties with value 0 should be omitted.

lorenzkuhn commented on src/learning/cyclegan.py:
```diff
@@ -0,0 +1,96 @@
+    gen_xy.apply(utils.init_weights_gaussian)
+    gen_yx.apply(utils.init_weights_gaussian)
+
+    optimizer_discr_x = optim.Adam(discr_x.parameters(), lr=0.002)
```
We'll want to parameterise the learning rate.

kklein replied:
Yes.

kklein commented on src/backend/static/styles/style.css:
```diff
@@ -0,0 +1,74 @@
+  display: flex;
+}
+
+left {
```
Why is left styling different from right styling?

kklein commented on src/backend/static/styles/style.css:
```diff
@@ -0,0 +1,74 @@
+  background: rgb(71, 70, 70);
+  box-shadow: inset 1px 1px 1px white;
+  color: rgb(0, 0, 0);
+  border: none;
```
Properties with irrelevant values should be omitted. I'm not 100% sure but I think this applies to this case.

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+<html>
+  <head>
+    <!-- Global site tag (gtag.js) - Google Analytics -->
+    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-45266465-2"></script>
```
Why load Analytics right at the beginning? Let's try to offer a better user experience by loading it last.

phil9987 replied:
I did this because google suggested to do so. Where would you suggest to put it, in the end of 'head'?

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+        <section>
+          <img id='uploaded_image' height='256' width='256'>
+        </section>
+        <form id='upload_form' action='' enctype='multipart/form-data' method='post'>
```
Can we remove 'action='''?

phil9987 replied:
good point, we can actually remove everything and only leave '< form >' because these are artifacts from when I used the form.submit functionality

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+            <input id='file_input' type="file" name="file" accept="image/*" required="true">
+          </div>
+        </form>
+        <p><strong>Status: </strong></p><p id='status_text'>Please select image to transform.</p>
```
Consider moving this styling to css.

phil9987 replied:
for this I would need to create a custom tag instead of < p >? or I might just omit it as it's not essential.

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-45266465-2"></script>
+    <script>
+      window.dataLayer = window.dataLayer || [];
+      function gtag(){dataLayer.push(arguments);}
```
Fix spacing.
Move into lambda function.

lorenzkuhn commented on src/learning/utils.py:
```diff
@@ -0,0 +1,51 @@
+        transforms.Resize(image_size),
+        transforms.CenterCrop(image_size),
+        transforms.ToTensor(),
+        transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])
```
Why are we setting the means and standard deviations to be 0.5 for all channels?

kklein replied:
Post hoc rationalization: We are more keen on discovering features relative to the image than relative to other images. In particular, we have no reason to treat channels differently.
Reason: Other people do it.
Does that answer your question?

lorenzkuhn replied:
Kind of 😁

lorenzkuhn commented on src/learning/utils.py:
```diff
@@ -0,0 +1,51 @@
+
+def init_weights_gaussian(model):
+    classname = model.__class__.__name__
+    if classname.find('Conv') != -1:
```
What happens to Conv Layers?

kklein replied:
Their weights are initialized according to a Gaussian distribution.
Other layers, such as ReLU or Normalization layers do not have weights that are meant to be initialized.

lorenzkuhn replied:
Ah, I see. 
I think normalization layers actually do have weights to initialise, i.e. the scaling factor and the offset.

kklein replied:
Very good point.
My understanding is that InstanceNorm2d is instantiated without learnable weights by default.

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+          }
+          return response;
+        }
+        // uploads an image file to the server, fetches the inferred image and displays it 
```
Exceeds line length.

kklein replied:
Remove trailing whitespace.

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+        // uploads an image file to the server, fetches the inferred image and displays it 
+        function upload(file) {
+          let uploaded_image = document.getElementById('uploaded_image');
+          let status_text = document.getElementById('status_text')
```
Add missing semicolon.

kklein replied:
I think this could be a const.

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+          data.append('file', file);
+          fetch(url, { method: 'POST',body: data})
+          .then(handleErrors)
+          .then(response => {
```
Wrap parameters in parens.

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+            }
+            return response.blob();
+          })
+          .then(blob => {
```
Wrap parameters in parens.

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+          fetch(url, { method: 'POST',body: data})
+          .then(handleErrors)
+          .then(response => {
+            if(!response.ok) { 
```
Remove trailing whitespace.

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+          .then(handleErrors)
+          .then(response => {
+            if(!response.ok) { 
+              throw response; 
```
Remove trailing whitespace.

lorenzkuhn commented on src/learning/cyclegan.py:
```diff
@@ -0,0 +1,96 @@
+    mse = nn.MSELoss(reduction='elementwise_mean').to(device)
+    loss_function = CycleGANLoss(regularizer, False, device)
+
+    for epoch_index in range(n_epochs):
```
We don't use epoch_index, so we could also discard it.
```suggestion
    for _ in range(n_epochs):
```

kklein replied:
Kept for logging, which will take place in the near future.

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+          .catch(function(error) {
+            console.log("error occurred");
+            console.log(error);
+            status_text.innerHTML = error;
```
Do you really have to use .innerHTML? This is usually frowned upon for security concerns.

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+          const input = document.getElementById('file_input');
+          input.addEventListener('change', () => upload(input.files[0]), false);
+        }
+        initialize();    
```
Remove trailing whitespace.

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+            status_text.innerHTML = error;
+          });
+        }
+        // sets up the app logic, declares required variables, contains all the other functions
```
Respect maximum line length.

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+        }
+        // uploads an image file to the server, fetches the inferred image and displays it 
+        function upload(file) {
+          let uploaded_image = document.getElementById('uploaded_image');
```
I think this could be a const. Independently of that, we could just overwrite this property without storing it in a variable, right?

phil9987 replied:
right

lorenzkuhn commented on src/learning/cyclegan.py:
```diff
@@ -0,0 +1,96 @@
+        train_data_y, batch_size=batchsize, shuffle=True, num_workers=4)
+
+    mse = nn.MSELoss(reduction='elementwise_mean').to(device)
+    loss_function = CycleGANLoss(regularizer, False, device)
```
Seems like we only ever create CycleGANLoss with inverted=False, so we might want to remove the inverted argument.

kklein replied:
We will instantiate it with inverted=True, i.e. for the discriminators, once we start logging.

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+          const data = new FormData();
+          data.append('file', file);
+          fetch(url, { method: 'POST',body: data})
+          .then(handleErrors)
```
Fix indentation.

phil9987 replied:
do you mean the 'then' should be indented?

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+            return response.blob();
+          })
+          .then(blob => {
+            let image = document.getElementById('inferred_image');
```
We could just overwrite this property without storing it in a variable, right?

phil9987 replied:
right

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+            let image = document.getElementById('inferred_image');
+            const objectUrl = URL.createObjectURL(blob);
+            image.src = objectUrl;
+            status_text.innerHTML = 'Here you go. Feel free to upload another image.'
```
Respect maximum line length.

kklein commented on src/backend/templates/index.html:
```diff
@@ -0,0 +1,85 @@
+            image.src = objectUrl;
+            status_text.innerHTML = 'Here you go. Feel free to upload another image.'
+          })
+          .catch(function(error) {
```
Replace by arrow function.

lorenzkuhn commented on src/learning/cyclegan.py:
```diff
@@ -0,0 +1,96 @@
+            gen_yx, device)
+        loss = loss_function(cycle_data)
+
+        loss.backward(retain_graph=True)
```
Let's discuss again why this is necessary.

lorenzkuhn commented on src/learning/utils.py:
```diff
@@ -0,0 +1,51 @@
+
+
+def log_loss(loss):
+    with open('log_loss.csv', 'a') as log_file:
```
Detail: Sound like this has something to do with the logarithm of the loss, not the log file for the loss.

kklein replied:
Done.

kklein committed:
Remove bias from convolutional layers following normalization.

kklein committed:
Rename modules file.

kklein committed:
Address review suggestions.

kklein committed:
Remove redundant environment.

kklein committed:
Parametrize learning rate.

kklein committed:
Rename function.

kklein committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenzkuhn committed:
Update READMe

lorenz-openai committed:
Rename requirements file

lorenz-openai committed:
Fix import order

phil9987 committed:
implement changes of code review

phil9987 committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenz-openai committed:
Remove secret key

lorenz-openai committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
code review

lorenz-openai committed:
Actually remove secret ke

phil9987 committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
fix Generator issue

lorenz-openai committed:
Refactor model and transform function loading

lorenz-openai committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
restructure get and post

phil9987 committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenz-openai committed:
Add comment about file size limit

lorenz-openai committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenz-openai committed:
Fix reference to is_allowed_file()

lorenz-openai committed:
Remove trailing white space

kklein committed:
Address css review comments.

lorenz-openai committed:
Remove trailing whitespace

kklein committed:
Merge conflict.

lorenz-openai committed:
Make extraction of image name explicit

lorenz-openai committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
fix comma bug in nn_modules, add codereview changes

phil9987 committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenz-openai committed:
Make file exentsion removal more robust

lorenz-openai committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
add utils in dockerfile

phil9987 committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

kklein committed:
Use mean instead of elementwise_mean keyword for loss functions.
This shouldn't lead to a functional difference. The latter option will simply be deprecated in future releases.

kklein committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenz-openai committed:
Restructure flask configuration

lorenz-openai committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

lorenz-openai committed:
Add flask config file

lorenz-openai committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
essential fixes, design downgraded

phil9987 committed:
uncommented model load

kklein committed:
Add parsing of arguments for memory usage experiments.

kklein committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

Commit:
Resolve merge conflict

kklein committed:
Fix typo.

kklein committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

kklein committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

Commit:
Resolve conflict

lorenzkuhn committed:
Update filenames to those used in cyclegan repo

lorenzkuhn committed:
Update README

kklein committed:
Implement adaptive learning rate.

kklein committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
nginx first try

phil9987 committed:
Merge branch 'master' of https://github.com/lorenzkuhn/CycleGanApp

phil9987 committed:
add nginx to env
